Allow fetching two different TMS APIs

jraddaoui opened the pull request:
- The TMS version can be selected using the `ATOM_DRMC_TMS_VERSION` env.
  variable. Allowed values are '1' or '2'. This values are not real TMS
  API versions, just a way to differentiate between the two current
  implementations.
- Reorganize TMS related classes to follow an abstract class that
  defines the required methods for all versions and a wrapper class
  that creates an instance of those classes based on the configuration,
  making it transparent to the controllers and tasks that use them.
- Merge creation and update of artwork records in a single method.
- Move components parent retrieval/creation to TMS classes and minor
  cosmetic changes in qtPackageExtractorMETSArchivematicaDIP.
- Fix year parsing in Elasticsearch documents.
- Improve ingestion activity endpoint SQL query to be able use local
  thumbnails (from v2) and external URLs (from v1).
- Add more LODs for component statuses.

jraddaoui commented on apps/qubit/config/qubitConfiguration.class.php:
```diff
@@ -169,5 +170,17 @@ protected function bootstrapDrmc()
+
+    // This are not real TMS API versions, just a way to differentiate
+    // between the two current implementations.
+    $allowedTmsVersions = array('1', '2');
```
I didn't want to use the client names in here, but I'm open to better suggestions than "1" and "2" ;)

djjuhasz replied:
I think this is fine.  The important bit is that we have a way to distinguish between them, and can iterate on the versions.

djjuhasz reviewed (approved):
Looks good to me! 👍

jraddaoui commented:
Thanks @djjuhasz! This is finally ready for another review after their latest updates to their TMS API.

djjuhasz commented on lib/task/binder/arUpdateArtworkWorker.class.php:
```diff
@@ -87,7 +87,7 @@ public function run($id)
     {
       // Update artwork
```
Nitpick: I think this comment can be removed as the code is self-explanatory.

djjuhasz commented on lib/task/binder/arUpdateCollectionTmsTask.class.php:
```diff
@@ -88,7 +88,7 @@ protected function execute($arguments = array(), $options = array())
       if ($needsUpdate)
       {
         // Update artwork
```
Nitpick: I think this comment can be removed as the code is self-explanatory.

djjuhasz commented on lib/task/binder/tms/arFetchTmsBase.class.php:
```diff
@@ -0,0 +1,291 @@
+
+  protected function updateRemainingEsDocs($artwork, $updatedIoIds)
+  {
+    // Update non already updated artwork descendant IOs
```
Nitpick: the wording of this comment is confusing.

djjuhasz commented on lib/task/binder/tms/arFetchTmsBase.class.php:
```diff
@@ -0,0 +1,291 @@
+  protected function getOrCreateComponentsParent($artwork)
+  {
+    $criteria = new Criteria;
+    $criteria->add(QubitInformationObject::PARENT_ID, $artwork->id);
```
Isn't this getting the _children_ of $artwork (e.g. information_object.parent_id = $artwork->id)?

jraddaoui replied:
By limiting to the `app_drmc_lod_description_id` LOD and using `QubitInformationObject::getOne` it should only get the intermediate IO (called "Components") and not the actual TMS component IOs.

djjuhasz replied:
Ah okay, so this is getting a "Components" information object that is a child of the the Artword IO, and a parent to the individual component IOs?

jraddaoui replied:
Yes, that's right!

djjuhasz replied:
👍

djjuhasz commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPluginUtil.class.php:
```diff
@@ -44,6 +44,22 @@ public static function convertDate($date)
+  public static function parseYear($dateString)
+  {
+    // Do not use date_parse if we already have a year
+    if (strlen($dateString) === 4 && preg_match("/\d{4}/", $dateString) === 1)
```
Should use [yoda conditions](https://en.wikipedia.org/wiki/Yoda_conditions)

djjuhasz replied:
Actually, I guess Yoda conditions aren't necessary because it's not comparing the variable directly with a constant?

djjuhasz commented on lib/task/binder/tms/arFetchTmsV1.class.php:
```diff
@@ -0,0 +1,476 @@
+      }
+    }
+
+    return null;
```
Huh, I haven't been using `return null` when a return value is normally expected, but you are doing it according to the standard: https://symfony.com/doc/3.4/contributing/code/standards.html  👍

djjuhasz commented on lib/task/binder/tms/arFetchTmsV1.class.php:
```diff
@@ -0,0 +1,476 @@
+    $criteria->add(QubitInformationObject::LFT, $artwork->lft, Criteria::GREATER_THAN);
+    $criteria->add(QubitInformationObject::RGT, $artwork->rgt, Criteria::LESS_THAN);
+    $criteria->add(QubitInformationObject::LEVEL_OF_DESCRIPTION_ID, $this->componentLevels, Criteria::IN);
+    foreach (QubitInformationObject::get($criteria) as $component)
```
I think a blank line above the foreach is better here.

djjuhasz commented on lib/task/binder/tms/arFetchTmsV1.class.php:
```diff
@@ -0,0 +1,476 @@
+      }
+      else
+      {
+        // Move childs to parent of the component
```
I don't understand why the children are being re-parented here.

djjuhasz replied:
Oh I see, the children are re-parented before deleting the component. I think it would helpful to note this in the comment.

djjuhasz commented on lib/task/binder/tms/arFetchTmsV1.class.php:
```diff
@@ -0,0 +1,476 @@
+              // Add property for each attribute
+              $count = 0;
+              $propertyName = $propertyValue = null;
+              foreach ($item as $key => $value)
```
Blank line before foreach.

djjuhasz commented on lib/task/binder/tms/arFetchTmsV2.class.php:
```diff
@@ -0,0 +1,430 @@
+
+          case 'terms':
+            $content = array();
+            foreach ($value as $term)
```
blank line before foreach

djjuhasz commented on lib/task/binder/tms/arFetchTmsV2.class.php:
```diff
@@ -0,0 +1,430 @@
+        $criteria = new Criteria;
+        $criteria->add(QubitInformationObject::IDENTIFIER, $componentData['componentId']);
+        $criteria->add(QubitInformationObject::PARENT_ID, $componentsParent->id);
+        if (null === $component = QubitInformationObject::getOne($criteria))
```
Blank line before control statement.

djjuhasz commented on lib/task/binder/tms/arFetchTmsV2.class.php:
```diff
@@ -0,0 +1,430 @@
+
+            case 'attributes':
+              $content = array();
+              foreach ($value as $attribute)
```
Blank line before control statement

djjuhasz commented on lib/task/binder/tms/arFetchTmsV2.class.php:
```diff
@@ -0,0 +1,430 @@
+
+            case 'textEntries':
+              $content = array();
+              foreach ($value as $textEntry)
```
Blank line before control statement

djjuhasz commented on lib/task/binder/tms/arFetchTmsV2.class.php:
```diff
@@ -0,0 +1,430 @@
+    $criteria->add(QubitInformationObject::RGT, $artwork->rgt, Criteria::LESS_THAN);
+    $criteria->add(QubitInformationObject::LEVEL_OF_DESCRIPTION_ID, $this->componentLevels, Criteria::IN);
+    $criteria->add(QubitInformationObject::ID, $tmsComponentsIoIds, Criteria::NOT_IN);
+    foreach (QubitInformationObject::get($criteria) as $oldComponent)
```
Blank line before control statement

djjuhasz commented on lib/task/binder/tms/arFetchTmsV2.class.php:
```diff
@@ -0,0 +1,430 @@
+    }
+  }
+
+  private function getTmsComponentsData($artwork, $componentsParent)
```
I like the decomposition of the `getTmsComponentsData` functionality into it's own method.  Is it possible to do the same for the V1 API to make it easier to compare the two?

jraddaoui replied:
I didn't try to normalize and abstract that part because the process to get the components data is quite different between versions:

In the v1, the TMS Component IDs are obtained from the TMS Object response and they are fetched one by one using those IDs. Whoever, in v2 the TMS Object response doesn't contain the TMS Components IDs and the only way to get them is in a single request which includes all the metadata for all of them.

Please, let me know if you still one to normalize a little more the process and I'll give it a try.

djjuhasz replied:
Ah okay, I think it's fine to leave this as it is then. :)

djjuhasz reviewed (changes requested):
@jraddaoui nice work on big PR!  I've requested a few changes and suggested a few tweaks.

jraddaoui commented:
Thanks @djjuhasz! Back to you.

djjuhasz reviewed (approved):
LGTM! 👍

jraddaoui committed:
Allow fetching two different TMS APIs

- The TMS version can be selected using the `ATOM_DRMC_TMS_VERSION` env.
  variable. Allowed values are '1' or '2'. This values are not real TMS
  API versions, just a way to differentiate between the two current
  implementations.
- Reorganize TMS related classes to follow an abstract class that
  defines the required methods for all versions and a wrapper class
  that creates an instance of those classes based on the configuration,
  making it transparent to the controllers and tasks that use them.
- Merge creation and update of artwork records in a single method.
- Move components parent retrieval/creation to TMS classes and minor
  cosmetic changes in qtPackageExtractorMETSArchivematicaDIP.
- Fix year parsing in Elasticsearch documents.
- Improve ingestion activity endpoint SQL query to be able use local
  thumbnails (from v2) and external URLs (from v1).
- Add more LODs for component statuses.
